### PR TITLE
Use read-only wrapper for codepoints in Mark class instead of making a full array copy

### PR DESCRIPTION
**Pull request by OptimumCode**
This PR should solve the performance degradation introduced with PR #182.

A proper benchmark in the project should prevent such issues in future (I think it would be better to add it in a separate PR).

@krzema12 @aSemy could you please take a look at the PR?

Relates to #204

### CODE REVIEW

File: src/commonMain/kotlin/it/krzeminski/snakeyaml/engine/kmp/scanner/StreamReader.kt [hunk 1]
```diff
@@ -98,7 +98,7 @@ class StreamReader(
             index = index,
             line = line,
             column = column,
-            codepoints = codePointsWindow.toList(),
+            codepoints = codePointsWindow.toReadOnlyListAdapter(),
```
Comment (aSemy) [inline src/commonMain/kotlin/it/krzeminski/snakeyaml/engine/kmp/scanner/StreamReader.kt:101]: There's also [`asList()`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/as-list.html). I've tried it and I think it works the same. WDYT?

Comment (OptimumCode) [reply src/commonMain/kotlin/it/krzeminski/snakeyaml/engine/kmp/scanner/StreamReader.kt:101]: Yea, indeed. Completely forgot about that

File: src/commonMain/kotlin/it/krzeminski/snakeyaml/engine/kmp/exceptions/Mark.kt [hunk 2]
```diff
@@ -39,7 +39,7 @@ class Mark @JvmOverloads constructor(
 ) {
 
     @Deprecated("Converted to immutable List<Int>, replace with `codepoints`")
-    val buffer: IntArray = codepoints.toIntArray()
+    val buffer: IntArray by lazy { codepoints.toIntArray() }
```
[the remark this answers is not in the document]
Comment (OptimumCode) [reply src/commonMain/kotlin/it/krzeminski/snakeyaml/engine/kmp/exceptions/Mark.kt:42]: I agree. Maybe also increase depreciation level to Error?

File: src/commonMain/kotlin/it/krzeminski/snakeyaml/engine/kmp/scanner/StreamReader.kt (hunk 1 above)
Comment (OptimumCode) [reply src/commonMain/kotlin/it/krzeminski/snakeyaml/engine/kmp/scanner/StreamReader.kt:101]: Interesting note though:

For the JS target the method `asList` returns a mutable list

```kotlin
public actual inline fun IntArray.asList(): List<Int> {
    return this.unsafeCast<Array<Int>>().asList()
}

public actual fun <T> Array<out T>.asList(): List<T> {
    return ArrayList<T>(this.unsafeCast<Array<Any?>>())
}
```

File: src/commonMain/kotlin/it/krzeminski/snakeyaml/engine/kmp/exceptions/Mark.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (aSemy) [reply src/commonMain/kotlin/it/krzeminski/snakeyaml/engine/kmp/exceptions/Mark.kt]: Thanks for changing it.

Maybe we can change the deprecation level to error later, but for now it's fine as a warning. AFAIK Kaml is the only user of SnakeKMP, so as long as Kaml updates, it's fine.

**Review by aSemy — APPROVED** [review]

**Comment by OptimumCode** [comment]
@aSemy Maybe we should remove the performance test since it results in different time execution on different platforms? And there is also #208 PR that should cover that

**Comment by aSemy** [comment]
I'm planning some more refactoring #177, so it'd be useful to have right now, even if it's not very strict. I've relaxed the timing, but yes, proper performance testing will be better.

**Comment by krzema12** [comment]
Thanks for taking care of this @OptimumCode @aSemy! Is it all we need for the patch release?

**Comment by OptimumCode** [comment]
Hi, @krzema12. I believe that should be all we need

**Comment by aSemy** [comment]
Yes, this PR fixes #204.

**Comment by krzema12** [comment]
Ack, I'll prepare the release tentatively on Monday.

**Pull request merged**